python/mxnet/cython/symbol.pyx

```diff
@@ -0,0 +1,226 @@
+
+import sys as _sys
+import ctypes as _ctypes
+from numbers import Number as _Number
+from ..name import NameManager
+from ..attribute import AttrScope
+from libcpp.vector cimport vector
+from libcpp.string cimport string
+from cpython.version cimport PY_MAJOR_VERSION
+
+include "./base.pyi"
+
+cdef extern from "nnvm/c_api.h":
+    const char* NNGetLastError();
+    int NNSymbolListAtomicSymbolCreators(nn_uint *out_size,
+                                         AtomicSymbolCreator **out_array);
+    int NNSymbolCreateAtomicSymbol(AtomicSymbolCreator creator,
+                                   nn_uint num_param,
+                                   const char **keys,
+                                   const char **vals,
+                                   SymbolHandle *out);
+    int NNSymbolFree(SymbolHandle symbol);
+    int NNSymbolSetAttrs(SymbolHandle symbol,
+                         nn_uint num_param,
+                         const char** keys,
+                         const char** values);
+    int NNSymbolCompose(SymbolHandle sym,
+                        const char* name,
+                        nn_uint num_args,
+                        const char** keys,
+                        SymbolHandle* args);
+
+cdef extern from "mxnet/c_api.h":
+    int MXSymbolGetAtomicSymbolInfo(AtomicSymbolCreator creator,
+                                    const char **name,
+                                    const char **description,
+                                    nn_uint *num_doc_args,
+                                    const char ***arg_names,
+                                    const char ***arg_type_infos,
+                                    const char ***arg_descriptions,
+                                    const char **key_var_args,
+                                    const char **return_type);
+
+cdef class SymbolBase:
+    """Symbol is symbolic graph."""
+    # handle for symbolic operator.
+    cdef SymbolHandle chandle
+
+    cdef _set_handle(self, handle):
+        cdef unsigned long ptr
+        if handle is None:
+            self.chandle = NULL
+        else:
+            ptr = handle.value
+            self.chandle = <SymbolHandle>(ptr)
+
+    property handle:
+        def __get__(self):
+            if self.chandle == NULL:
+                return None
+            else:
+                return _ctypes.cast(<unsigned long>self.chandle, _ctypes.c_void_p)
+        def __set__(self, value):
+            self._set_handle(value)
+
+    def __init__(self, handle):
+        self._set_handle(handle)
+
+    def __dealloc__(self):
+        CALL(NNSymbolFree(self.chandle))
+
+    def _set_attr(self, **kwargs):
+        """Set the attribute of the symbol.
+
+        Parameters
+        ----------
+        **kwargs
+            The attributes to set
+        """
+        SymbolSetAttr(self.chandle, kwargs)
+
+    def __reduce__(self):
+        return (_symbol_cls, (None,), self.__getstate__())
+
```

Comment:
I decided that compose can in slow path and moved it to common area, can be moved back later